### PR TITLE
Force upsample  to be float32

### LINKED ISSUE #121072: Correctly handle `F.interpolate` upsample with amp
### 🐛 Describe the bug

https://github.com/pytorch/pytorch/issues/104157

As this bug seems to be still a quite tricky  for many user can we let correctly handle this case when using amp?

I see developer are using manual tricks to workaround this bug:
https://github.com/huggingface/transformers/pull/28504/files#r1455033425

But as often amp is automated on the whole model it could be insidious for many user.

Can amp automatically handle this case internally? 

### Versions

main/nightly

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @mcarilli @ptrblck @leslie-fang-intel @jgong5

### PR DESCRIPTION
**Pull request by bhack**
Fixes #121072

BC-breaking note:
This is not technically bc-breaking any behavior but will lead to an expected significant performance change for amp + deterministic.
cc @zou3519 any way we can tell people how to register a fallthrough for that key to recover the old behavior if they want?

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: aten/src/ATen/autocast_mode.cpp [hunk 1]
```diff
@@ -330,6 +330,8 @@ TORCH_LIBRARY_IMPL(aten, Autocast, m) {
   KERNEL_CUDA(cdist, fp32)
   KERNEL_CUDA(renorm, fp32)
   KERNEL_CUDA(logsumexp, fp32)
+  KERNEL_CUDA(upsample, fp32)
+  KERNEL_CUDA2(upsample, fp32)
```
Comment (albanD) [inline aten/src/ATen/autocast_mode.cpp:334]: Id on't think you need this one. This is for when you want to specify a specific overload.

Comment (bhack) [reply aten/src/ATen/autocast_mode.cpp:334]: This one? Do you mean CUDA?

Comment (albanD) [reply aten/src/ATen/autocast_mode.cpp:334]: The CUDA2 yes. I expect this is not going to compile

Comment (bhack) [reply aten/src/ATen/autocast_mode.cpp]: Done

**Review by albanD — COMMENTED** [review]
Let's see what CI thinks, but we should also add a test for this!

**Review by ezyang — DISMISSED** [review]

**Comment by bhack** [comment]
About the user facing doc I don't see that we had upsample in the list right?
https://github.com/pytorch/pytorch/blob/main/docs%2Fsource%2Famp.rst

Is thus probably cause that list is not auto-generated and it is out of sync?

**Comment by bhack** [comment]
Green lights here.

**Comment by bhack** [comment]
Are these errors related to this PR?

**Comment by nWEIdia** [comment]
> ImportError: /opt/conda/envs/py_3.8/lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/lib/libtorch_python.so)
The error seems to be build env issues.

**Comment by nWEIdia** [comment]
> @nWEIdia Yes it is not related but I was just asking if you have a specific opinion at Nvidia about mixed precision float16 or bfloat in upsample kernels related to real model impact (see #121072).

Does this fix convergence issues of certain models? Has any perf benchmarking study been performed?  We could keep an eye on torch bench results.
cc @xuzhao9  @eqy @ptrblck

**Comment by nWEIdia** [comment]
hud.pytorch.org signals do not look good, this PR seems to have caused UNet regressions. 
cc @huydhn @clee2000

**Comment by bhack** [comment]
> hud.pytorch.org signals do not look good, this PR seems to have caused UNet regressions.

Inference or training?

**Comment by huydhn** [comment]
During training accuracy check I think https://github.com/pytorch/pytorch/actions/runs/8292947312/job/22696390483#step:18:3637

**Comment by bhack** [comment]
In the CI log I see only `allclose not within tol=0` that it could be ambiguous. Can we see the real and expect values?

**Comment by huydhn** [comment]
May be @desertfire or @anijain2305 could help here?

**Comment by bhack** [comment]
Where  can we retrieve the current and the expected values?

**Comment by ezyang** [comment]
Assuming you can get a working local torchbench setup, you should be able to run the benchmarker script to reproduce the problem, then you can probe values as needed.

**Comment by bhack** [comment]
It will be a little bit complicated honestly as other than torchbemch I need to fully recompile torch locally as this is a c++ level PR.

Do you have some global shared build cache at META to just recompile it quickly?

The  check that it is failing in the test is only for exact reproducibility (that I suppose it cannot be expected with this PR) but it doesn't tell us if its numbers are going better or worse.

**Comment by eqy** [comment]
The unet test is failing with `eager_two_runs_differ` which seems to suggest that this PR is introducing nondeterministic behavior in eager where there previously was not, so you might want to attempt reproducing that.

**Comment by bhack** [comment]
> The unet test is failing with eager_two_runs_differ which seems to suggest that this PR is introducing nondeterministic behavior in eager where there previously was not, so you might want to attempt reproducing that.

But it is a `dynamo` specific test right? We are not touching `dynamo` directly so how we can check the dynamo handling of this?

```python
2024-03-15T10:42:02.6025349Z Error: 1 models have accuracy status regressed:
2024-03-15T10:42:02.6025979Z     pytorch_unet
2024-03-15T10:42:02.6026251Z 
2024-03-15T10:42:02.6026270Z 
2024-03-15T10:42:02.6028421Z If this change is expected, you can update `benchmarks/dynamo/ci_expected_accuracy/inductor_torchbench_training.csv` to reflect the new baseline.
2024-03-15T10:42:02.6029396Z from pytorch/pytorch root, run
2024-03-15T10:42:02.6030145Z `python benchmarks/dynamo/ci_expected_accuracy/update_expected.py 2770e3addd9f05101705f0fef85a163e0034b8a5`
2024-03-15T10:42:02.6031110Z and then `git add` the resulting local changes to expected CSVs to your commit.
2024-03-15T10:42:02.6031618Z 
2024-03-15T10:42:02.6522675Z ##[error]Process completed with exit code 1.
2024-03-15T10:42:02.6597430Z Prepare all required actions
```

**Comment by bhack** [comment]
Is it related to https://github.com/pytorch/pytorch/issues/76176?

As near the error I see a lot of of dynamo interpolate log message like:
```python
2024-03-15T10:39:05.7357669Z skipping cudagraphs due to deterministic index put. Found from : 
2024-03-15T10:39:05.7359294Z    File "/var/lib/jenkins/.local/lib/python3.10/site-packages/torchvision/models/detection/roi_heads.py", line 415, in torch_dynamo_resume_in_paste_mask_in_image_at_407
2024-03-15T10:39:05.7360576Z     mask = F.interpolate(mask, size=(h, w), mode="bilinear", align_corners=False)
```

**Comment by ezyang** [comment]
Oh, this is unfortunate... do we really not have a deterministic float32 upsample implementation.... @kurtamohler do you know?

**Comment by bhack** [comment]
I see:
https://github.com/pytorch/pytorch/blob/ae913175c30c13f5ecea499b715f26f83d5251ce/aten/src/ATen/native/cuda/UpSampleBicubic2d.cu#L293

**Comment by ezyang** [comment]
This is really dorky, I hate to ask you to do it, but... can we have a knob to choose if upsample is float32 or float16? Then we can force float16 in the Dynamo test suite which would resolve the problem

**Comment by bhack** [comment]
But we already have a lot of other networks in benchmark where deterministic is set to false in the config metadata.
E.g. [stable diffusion-unet](https://github.com/pytorch/benchmark/blob/2196021e9bc0b72a547121bbf298ae854a85a21a/torchbenchmark/models/stable_diffusion_unet/metadata.yaml#L5)
Does it make sense to set unet deterministic false?

**Comment by bhack** [comment]
I've tried to add this PR to the benchmark repo as the `atomicAdd` it seems (at least by inline comments) affecting only some backward kernels:
https://github.com/pytorch/benchmark/pull/2204

**Comment by bhack** [comment]
I think it is mainly:
https://github.com/pytorch/pytorch/blob/0b68a28c87df2c6eb2cf530be4659b5a2f8a95b0/aten/src/ATen/native/cuda/UpSample.cuh#L214-L216

I suppose that our issue it could be similar to:
https://github.com/NVIDIA/framework-reproducibility/blob/master/doc/d9m/tensorflow_status.md?plain=1#L590-L593

So I suppose why it is going to emerge only with `flaot32` backwards.

**Comment by bhack** [comment]
See also:
https://github.com/NVIDIA/framework-reproducibility/issues/18#issuecomment-663266283

**Comment by kurtamohler** [comment]
> Oh, this is unfortunate... do we really not have a deterministic float32 upsample implementation.... @kurtamohler do you know?

Not for backward CUDA. This is in the docs under `torch.nn.functional.interpolate` on [this page](https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch-use-deterministic-algorithms:~:text=torch.nn.functional.interpolate,trilinear). But I suppose we should have `torch.nn.Upsample` listed there as well

**Comment by ezyang** [comment]
I think we could potentially torch.compile a deterministic version of upsample, using the same playbook from https://dev-discuss.pytorch.org/t/a-pure-python-implementation-of-roi-align-that-looks-just-like-its-cuda-kernel/1266 We would do the forwards and then automatically differentiate it to get backwards.

This is the biggest of yak shaving, but would you be willing to give it a try @bhack ? I looked at bilinear2d (I'm not sure which sampling mode you actually need) and it looks to me that first class dims should work to directly transcribe this.

```
__global__ void upsample_bilinear2d_out_frame(
    const int n,
    const accscalar_t rheight,
    const accscalar_t rwidth,
    const bool align_corners,
    const PackedTensorAccessor<const scalar_t, 4> idata,
    PackedTensorAccessor<scalar_t, 4> odata) {
  int index = threadIdx.x + blockIdx.x * blockDim.x;
    
  const int batchsize = idata.size(0);
  const int channels = idata.size(1);
  const int height1 = idata.size(2);
  const int width1 = idata.size(3);
  const int width2 = odata.size(3);
    
  if (index < n) {
    const int w2 = index % width2; // 0:width2-1
    const int h2 = index / width2; // 0:height2-1
  
    const accscalar_t h1r = area_pixel_compute_source_index<accscalar_t>(
        rheight, h2, align_corners, /*cubic=*/false);
    const int h1 = h1r;
    const int h1p = (h1 < height1 - 1) ? 1 : 0;
    const accscalar_t h1lambda = h1r - h1;
    const accscalar_t h0lambda = static_cast<accscalar_t>(1) - h1lambda;
    //  
    const accscalar_t w1r = area_pixel_compute_source_index<accscalar_t>(
        rwidth, w2, align_corners, /*cubic=*/false);
    const int w1 = w1r;
    const int w1p = (w1 < width1 - 1) ? 1 : 0;
    const accscalar_t w1lambda = w1r - w1;
    const accscalar_t w0lambda = static_cast<accscalar_t>(1) - w1lambda;
    //  
    for (int n = 0; n < batchsize; n++) {
      for (int c = 0; c < channels; ++c) {
        const accscalar_t val = h0lambda *
                (w0lambda * idata[n][c][h1][w1] +
                 w1lambda * idata[n][c][h1][w1 + w1p]) +
            h1lambda *
                (w0lambda * idata[n][c][h1 + h1p][w1] +
                 w1lambda * idata[n][c][h1 + h1p][w1 + w1p]);
        odata[n][c][h2][w2] = static_cast<scalar_t>(val);
      } 
    }   
  }   
}  
```

**Comment by ezyang** [comment]
OK it turns out we already have upsample decomps e.g. https://github.com/pytorch/pytorch/pull/104182 and these should be differentiable. So all we have to do is torch.compile the computed backwards and that will be deterministic in this case. cc @lezcano @vfdev-5 @isuruf

**Comment by bhack** [comment]
> So all we have to do is torch.compile the computed backwards

Do we need to do it on unet benchmark repo?

**Comment by vfdev-5** [comment]
Decomposition for upsample bilinear2d is already autograd composite implicit: 

https://github.com/pytorch/pytorch/blob/2b4173e0def678b04efea46a85909bb86f7c3fd6/torch/_decomp/decompositions.py#L3526-L3537

**Comment by bhack** [comment]
It is still not clear to me if we need to compile the unet pytorch/benchmark or not.

Also, I don't know if this test it is correct with nightly as it seems we are cover by `decompositions`:

```python
import torch
from torch.nn.functional import interpolate

class InterpolateGradComputer(torch.nn.Module):
    def __init__(self, dtype, size):
        super().__init__()
        self.dtype = dtype
        self.size = size

    @torch.compile
    def forward(self, input):
        output = interpolate(input, size=self.size, mode='bicubic', align_corners=True)
        return output

def get_grads(dtype, size):
    input = torch.randn((2, 2, 1, 1), dtype=dtype, requires_grad=True).cuda()
    computer = InterpolateGradComputer(dtype, size).cuda()

    output = computer(input)

    grad_outputs = [torch.ones_like(output)]
    grads = torch.autograd.grad([output], [input], grad_outputs)
    return grads[0]

# Example usage
print(get_grads(torch.float32, (128, 128)).flatten())
print(get_grads(torch.float16, (128, 128)).flatten())
print(get_grads(torch.bfloat16, (128, 128)).flatten())

print(get_grads(torch.float32, (64, 64)).flatten())
print(get_grads(torch.float16, (64, 64)).flatten())
print(get_grads(torch.bfloat16, (64, 64)).flatten())

print(get_grads(torch.float32, (32, 32)).flatten())
print(get_grads(torch.float16, (32, 32)).flatten())
print(get_grads(torch.bfloat16, (32, 32)).flatten())
```

So how we are going to differentiate AMP with and without compilation?

**Comment by bhack** [comment]
What I meant about the example at https://github.com/pytorch/pytorch/pull/121324#issuecomment-2020470028

With compiler decorator:
```python
tensor([16384., 16384., 16384., 16384.], device='cuda:0')
tensor([16384., 16384., 16384., 16384.], device='cuda:0', dtype=torch.float16)
tensor([16384., 16384., 16384., 16384.], device='cuda:0', dtype=torch.bfloat16)
tensor([4096., 4096., 4096., 4096.], device='cuda:0')
tensor([4096., 4096., 4096., 4096.], device='cuda:0', dtype=torch.float16)
tensor([4096., 4096., 4096., 4096.], device='cuda:0', dtype=torch.bfloat16)
tensor([1024., 1024., 1024., 1024.], device='cuda:0')
tensor([1024., 1024., 1024., 1024.], device='cuda:0', dtype=torch.float16)
tensor([1024., 1024., 1024., 1024.], device='cuda:0', dtype=torch.bfloat16)
```
Without compiler decorator (eager):
```python
tensor([16384., 16384., 16384., 16384.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([256., 256., 256., 256.], device='cuda:0', dtype=torch.bfloat16)
tensor([4096., 4096., 4096., 4096.], device='cuda:0')
tensor([2048., 2048., 2048., 2048.], device='cuda:0', dtype=torch.float16)
tensor([256., 256., 256., 256.], device='cuda:0', dtype=torch.bfloat16)
tensor([1024., 1024., 1024., 1024.], device='cuda:0')
tensor([1024., 1024., 1024., 1024.], device='cuda:0', dtype=torch.float16)
tensor([256., 256., 256., 256.], device='cuda:0', dtype=torch.bfloat16)
```

**Comment by lezcano** [comment]
The triton generated backward is also going to have atomic adds, so how do you expect it to be more deterministic than our hand-written kernel @ezyang ?

**Comment by ezyang** [comment]
xref https://github.com/pytorch/pytorch/pull/121769 have convo only one place

**Comment by bhack** [comment]
But still `index_put` right.

Also here is more about forcing `fp32` in eager vs compile mode for AMP

**Comment by bhack** [comment]
The deterministic part of the problem is more on the benchmark https://github.com/pytorch/benchmark/pull/2204

**Comment by bhack** [comment]
So the big question here:

- Is the manual example results at https://github.com/pytorch/pytorch/pull/121324#issuecomment-2021411957 not covering enough real use cases?
- In the case that test input is enough for the backwards to be lower precision safe in compiled mode do we have a conditional AMP machinery in the pytorch core to differentiate between eager and compiled?

**Comment by bhack** [comment]
In the case instead with compiled generated code we are going to bypass `upsample_<whatever>` kernel I think this could be ready after the benchmark after this is merged https://github.com/pytorch/benchmark/pull/2204

File: benchmarks/dynamo/torchbench.yaml [hunk 2]
```diff
@@ -252,6 +252,7 @@ accuracy:
     eager_not_deterministic:
       # Models that deterministic algorithms can not be turned on for eager mode.
       - Background_Matting
+      - pytorch_unet
```
Comment (bhack) [inline benchmarks/dynamo/torchbench.yaml:255]: @ezyang @xuzhao9  Let me know what we want to do instead in the benchmark repo:
https://github.com/pytorch/benchmark/pull/2204#discussion_r1547637515

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -3193,7 +3193,7 @@ def _unsafe_index_put_(self, indices, values, accumulate=False):
 
 def needs_fallback_due_to_atomic_add_limitations(dtype):
     # tl.atomic_add does NOT support the following types
-    return dtype in {torch.int64, torch.bool, torch.bfloat16}
+    return dtype in {torch.int64, torch.bool, torch.bfloat16, torch.float16}
```
Comment (bhack) [inline torch/_inductor/lowering.py:3196]: @lezcano Can you review this?

**Review by lezcano — DISMISSED** [review]
The atomic_add fallback does not belong to this PR.

Even more, why would you make it fallback when it currently works?

```python
import torch


@torch.compile
def fn(x, y):
    return x.index_put_((torch.arange(1, device="cuda"),), y, accumulate=True)

x = torch.randn(3, dtype=torch.float16, device="cuda")
y = torch.randn(1, dtype=torch.float16, device="cuda")
fn(x, y)
```

**Comment by bhack** [comment]
> ```python
> import torch
> 
> 
> @torch.compile
> def fn(x, y):
>     return x.index_put_((torch.arange(1, device="cuda"),), y, accumulate=True)
> 
> x = torch.randn(3, dtype=torch.float16, device="cuda")
> y = torch.randn(1, dtype=torch.float16, device="cuda")
> fn(x, y)
> ```

ok probably there is some internal casting as I don't see a `float16` triton kenrel.

**Comment by bhack** [comment]
Green lights here.

File: benchmarks/dynamo/ci_expected_accuracy/dynamic_inductor_torchbench_training.csv [hunk 4]
```diff
@@ -194,7 +194,7 @@ pytorch_stargan,pass,6
 
 
 
-pytorch_unet,pass,7
+pytorch_unet,pass_due_to_skip,7
```
[the remark this answers is not in the document]
Comment (bhack) [reply benchmarks/dynamo/ci_expected_accuracy/dynamic_inductor_torchbench_training.csv:197]: It is expected also here. I don't know if by the change we have done to `eager_not_deterministic:` or to auto sync of `pytorch/benchmark`

[the remark this answers is not in the document]
Comment (lezcano) [reply benchmarks/dynamo/ci_expected_accuracy/dynamic_inductor_torchbench_training.csv:197]: Can you dig into that?

File: benchmarks/dynamo/torchbench.yaml [hunk 5]
```diff
@@ -252,6 +252,7 @@ accuracy:
     eager_not_deterministic:
       # Models that deterministic algorithms can not be turned on for eager mode.
       - Background_Matting
+      #- pytorch_unet
```
Comment (bhack) [inline benchmarks/dynamo/torchbench.yaml:255]: @lezcano Check the new test results as I have just commented out this

Comment (bhack) [reply benchmarks/dynamo/torchbench.yaml:255]: @lezcano It is done so you can check yourself. If we don't enable `pytorch_unet` under this section we have `eager_two_runs_differ`. 
If we add it we need to switch the expected to `pass_due_to_skip`. This is why we have changed all these csv to `pass_due_to_skip`.

Comment (bhack) [reply benchmarks/dynamo/torchbench.yaml:255]: Is it correct that this is going to impact both eager and non-eager expected results also if it is called `eager_not_deterministic`?

Comment (lezcano) [reply benchmarks/dynamo/torchbench.yaml:255]: This feels a bit annoying, but it also makes sense. Don't we have a list where we can bump the tolerance for a given model? WDYT @ezyang

**Comment by ezyang** [comment]
I really want to land this but I have lost the thread on what the plan is here. In particular, it seems CI is failing on this.

**Comment by bhack** [comment]
@ezyang If I have understood the thread correctly, we are not deterministic in eager mode and deterministic in compiled mode falling back in eager mode with sort if deterministic is required (right?).  But if we are going to make this change https://github.com/pytorch/pytorch/pull/121324#pullrequestreview-1976246727 it is seems that we are going to change also the non-eager expected values for tests. If you see the comment is eager but it seems not isolated to eager tests.

**Comment by ezyang** [comment]
Oh, I see. Yes, I agree that the reclassification is right. Go ahead and uncomment the line.

**Comment by bhack** [comment]
Done. 
What do you think about the compiled results vs non compiled one at https://github.com/pytorch/pytorch/pull/121324#issuecomment-2021411957 ?

**Comment by ezyang** [comment]
The difference does seem severe, although this is low precision. The gold standard is comparing against fp64 results; if we improve accuracy when compiling, we usually let the difference slide.

**Comment by bhack** [comment]
Or the compiled codepath it is not affected by this PR as it is not going to use this op and it could run float16/bfloat16 with AMP ?

**Comment by bhack** [comment]
As this section of code was totally refactored 5 days ago by @guangyey https://github.com/pytorch/pytorch/pull/124051 let me know what you wan to do with this

**Comment by bhack** [comment]
Also we have this one on the queue https://github.com/pytorch/pytorch/pull/124359 under review by @albanD

**Comment by albanD** [comment]
Ho yes sorry for the heavy refactoring. This was done so we can re-use these rules for non-cuda backends.
You can preserve the behavior you have here by adding these upsample methods to the list here https://github.com/pytorch/pytorch/blob/302d7e9a6ecc0d8e162f6d4ff8d067d7ba5bf4eb/aten/src/ATen/autocast_mode.cpp#L326

**Comment by bhack** [comment]
Can we clarify if it is ok for the compiled Vs eager code path?

**Comment by albanD** [comment]
This refactor will preserve the current behavior. So that won't change the current conclusion you reached above.

**Comment by bhack** [comment]
Thanks @albanD the question was general not only about the refactoring as it is not clear if these ops are still impacted by the amp float32 enforcing machinery when lowered/compiled.

**Comment by bhack** [comment]
@albanD It was adapted to the mentioned refactoring.

File: aten/src/ATen/autocast_mode.cpp [hunk 6]
```diff
@@ -239,7 +239,19 @@ Explicit registration for out-of-place ops
...
+  _(logsumexp)                        \
+  _(upsample_nearest1d)               \
+  _(_upsample_nearest_exact1d)        \
+  _(upsample_nearest2d)               \
+  _(_upsample_nearest_exact2d)        \
+  _(upsample_nearest3d)               \
+  _(_upsample_nearest_exact3d)        \
+  _(upsample_linear1d)                \
+  _(upsample_bilinear2d)              \
+  _(_upsample_bilinear2d_aa)          \
```
Comment (albanD) [inline aten/src/ATen/autocast_mode.cpp:251]: You have a whitespace at the end of this line that makes build very unhappy haha

Comment (bhack) [reply aten/src/ATen/autocast_mode.cpp]: True

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by xuzhao9** [comment]
This diff regresses torchbench pytorch_unet by 10%: https://github.com/pytorch/benchmark/issues/2247

**Comment by albanD** [comment]
Regressed in what sense? Speed or numerical?

**Comment by albanD** [comment]
I guess this is expected behavior given that we now upcast this op. cc @ezyang is that a regression that is problematic?

**Comment by ezyang** [comment]
This is an explicit correctness/speed tradeoff here. @bhack, is the memory usage regression you were expecting to see in this case?

**Comment by bhack** [comment]
A regression is expected for sure as we are working at `float32` also with amp on the upsampling.
My points were: 
- Are we forcing this also when compiled?
-  is the lowering going to let us to still use `flaot16/bfloat16`? 

Cause the gradient seems better at lower precision with the compiled code. 
See https://github.com/pytorch/pytorch/pull/121324#issuecomment-2021411957

**Pull request closed without merging**